documents/development/events.md

```diff
@@ -0,0 +1,108 @@
+
+### `eventData` table
+
+This is a “has-many” data table logging the data properties of each event. Each event can have 0 or many rows in this table.
+
+- _guid_ - Set by the Grouparoo server. The primary Key for this portion of an event’s data. Generated by the Grouparoo server. Should /not/ be auto-incremented but others unique (we use UUIDv4) + a `edt_` prefix.
+- _key_- Set by the client, required. The named key of this portion of an event’s data.
+- _value_- Set by the client, required. The value of this portion of an event’s data. All data is stored as strings.
+
+### `apiKeys` table
+
+The `apiKeys` table is stored within Grouparoo’s database, regardless of the `EventBackend` in use. Event’s must be created with an an `apiKey` which is authorized to create events.
+
+## Workflow
+
+1. Create an `apiKey` (`_key` in our example) which is authorized to create events. We will give it the name “Web API”.
+2. A Client clicks on a Google add for our website and lands on our homepage.
+3. Since we know nothing about this user, we’ll create an `anonymousId=person_1` and send an event about the page view:
+
+```js
+Grouparoo.init({
+  apiKey: "_key",
+  endpoint: "https://your-grouparoo-instance.company.com", // likely pre-set by the client library
+});
+
+Gropuaroo.track({
+  anonymousId: "person_1",
+  occurredAt: new Date(), // likley a default in the Client Library
+  type: "pageview",
+  data: {
+    path: "/",
+    utm_source: "google",
+    utm_medium: "cpc",
+    utm_campaign: "my-ad-campaign",
+    // ...
+  },
+});
+```
+
+Grouparoo will create a new `Profile` to track all the activity we can for “person_1”, and attempt to decorate their profile with any additional information we can load from your `Profile Property Rules`.
+
+Eventually, after visiting a few more pages, the user will sign in, and we’ll know their `userId`
+
+```js
+Gropuaroo.track({
+  userId: 123,
+  occurredAt: new Date(),
+  type: "sign-in",
+  data: {
+    via: "google-sign-in",
+    // ...
+  },
+});
+```
+
+From this point forward, we don’t need to send `anonymousId`. On the backend, Grouparoo will now be able to check if user 123 already has a `Profile`, and merge it with the profile we created for “person_1”. In this way, we will have a combined history for all the sessions of this user. If the same user signs in via another client / browser / phone, we will continue to create temporary profiles with an `anonymousId`, and eventually merge the profiles (and events!) once we have enough data.
+
+## Creating events outside of Grouparoo
+
+Grouparoo will not only process events created though our APIs, but will also periodically scan the events table for new events that have been added by an external system. This is a helpful if you want to write events directly from your backend applications. Grouparoo will look for events lacking a `profileGuid` and associate the events, making new `Profile`s as needed.
```

Comment:
It would be good to note that the schema should match and that we will write back the `profileGuid`.